app/forms/otp_verification_form.rb

```diff
@@ -15,9 +28,24 @@ def submit
 
   attr_reader :code, :user
 
-  def valid_direct_otp_code?
-    return false unless code.match? pattern_matching_otp_code_format
-    user.authenticate_direct_otp(code)
+  def validate_code_matches_format
+    return if code.match?(pattern_matching_otp_code_format)
+    errors.add(:code, :code_pattern_mismatch, type: :code_pattern_mismatch)
+  end
+
+  def validate_user_otp_presence
+    return if user.direct_otp.present?
+    errors.add(:code, :user_otp_missing, type: :user_otp_missing)
+  end
+
+  def validate_user_otp_expiration
+    return if !otp_expired?
+    errors.add(:code, :user_otp_expired, type: :user_otp_expired)
+  end
+
+  def validate_code_equals_user_otp
+    return if user.direct_otp == Base32::Crockford.normalize(code)
```

Comment:
I realize it was doing a direct comparison before, but WDYT of adding a secure_compare for extra security?
```suggestion
    return if ActiveSupport::SecurityUtils.secure_compare(user.direct_otp, Base32::Crockford.normalize(code))
```

Reply:
Included in 383b435.